File: backend/app/core/security.py
"""Password hashing and JWT creation/verification."""
from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from jose import JWTError, jwt
from passlib.context import CryptContext

from app.core.config import settings
from app.core.exceptions import UnauthorizedError

_pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

ACCESS = "access"
REFRESH = "refresh"


def hash_password(password: str) -> str:
    return _pwd_context.hash(password)


def verify_password(plain: str, hashed: str) -> bool:
    return _pwd_context.verify(plain, hashed)


def _now() -> datetime:
    return datetime.now(timezone.utc)


def create_access_token(*, user_id: str, role: str, email: str) -> str:
    expire = _now() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    payload = {
        "sub": user_id,
        "role": role,
        "email": email,
        "type": ACCESS,
        "iat": _now(),
        "exp": expire,
    }
    return jwt.encode(payload, settings.SECRET_KEY, algorithm=settings.ALGORITHM)


def create_refresh_token(*, user_id: str) -> tuple[str, str, datetime]:
    """Returns (token, jti, expires_at) so the session can be persisted."""
    jti = str(uuid.uuid4())
    expire = _now() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    payload = {"sub": user_id, "jti": jti, "type": REFRESH, "iat": _now(), "exp": expire}
    token = jwt.encode(payload, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return token, jti, expire


def decode_token(token: str, *, expected_type: str | None = None) -> dict[str, Any]:
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
    except JWTError as exc:
        raise UnauthorizedError("Invalid or expired token.") from exc
    if expected_type and payload.get("type") != expected_type:
        raise UnauthorizedError("Incorrect token type.")
    return payload

File: backend/src/utils/security.js
"use strict";

/**
 * Password hashing (bcrypt) and JWT creation/verification.
 * Mirrors app/core/security.py.
 *
 * bcryptjs is wire-compatible with the original passlib/bcrypt hashes, so
 * existing user password hashes continue to verify after the migration.
 */

const crypto = require("crypto");
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const { settings } = require("../config/env");
const { UnauthorizedError } = require("./errors");

const ACCESS = "access";
const REFRESH = "refresh";
const BCRYPT_ROUNDS = 12;

async function hashPassword(password) {
  return bcrypt.hash(password, BCRYPT_ROUNDS);
}

async function verifyPassword(plain, hashed) {
  try {
    return await bcrypt.compare(plain, hashed);
  } catch {
    return false;
  }
}

function createAccessToken({ userId, role, email }) {
  const payload = { sub: userId, role, email, type: ACCESS };
  return jwt.sign(payload, settings.SECRET_KEY, {
    algorithm: settings.ALGORITHM,
    expiresIn: `${settings.ACCESS_TOKEN_EXPIRE_MINUTES}m`,
  });
}

/** Returns { token, jti, expiresAt } so the session can be persisted. */
function createRefreshToken({ userId }) {
  const jti = crypto.randomUUID();
  const expiresInDays = settings.REFRESH_TOKEN_EXPIRE_DAYS;
  const payload = { sub: userId, jti, type: REFRESH };
  const token = jwt.sign(payload, settings.SECRET_KEY, {
    algorithm: settings.ALGORITHM,
    expiresIn: `${expiresInDays}d`,
  });
  const expiresAt = new Date(Date.now() + expiresInDays * 24 * 60 * 60 * 1000);
  return { token, jti, expiresAt };
}

function decodeToken(token, { expectedType = null } = {}) {
  let payload;
  try {
    payload = jwt.verify(token, settings.SECRET_KEY, { algorithms: [settings.ALGORITHM] });
  } catch {
    throw new UnauthorizedError("Invalid or expired token.");
  }
  if (expectedType && payload.type !== expectedType) {
    throw new UnauthorizedError("Incorrect token type.");
  }
  return payload;
}

module.exports = {
  ACCESS,
  REFRESH,
  hashPassword,
  verifyPassword,
  createAccessToken,
  createRefreshToken,
  decodeToken,
};
